### PR TITLE
Create markdown docs from docstring in py files

### PR DESCRIPTION
**Pull request by s-martin**
* use [pydoc-markdown](https://github.com/NiklasRosenstein/pydoc-markdown) to create markdown docs from docstrings in py files; generated docs can be found here: https://github.com/s-martin/RPi-Jukebox-RFID/blob/future3/pydoc-markdown/documentation/developers/docstring/README.md
* added to precommit hook to run, when py files are changed.
* changes for markdown formatting, e.g.
  * links
  * Code samples
  * Indentation
  * etc.

Pro/con evaluation of different tools can be found here https://github.com/s-martin/RPi-Jukebox-RFID/pull/31

Replaces #2159

### CODE REVIEW

**Comment by s-martin** [comment]
Found a problem we need to address, see https://github.com/NiklasRosenstein/pydoc-markdown/issues/308

**Comment by s-martin** [comment]
Enforcing the docstring style to sphinx did help for the code sample formatting problem.

**Comment by s-martin** [comment]
> Enforcing the docstring style to sphinx did help for the code sample formatting problem.

This also means that in the docstring we need to follow sphinx style (including rst for code samples), but for other formatting and links use markdown syntax.

**Comment by s-martin** [comment]
Fixes also #1771

**Comment by s-martin** [comment]
Could be merged, any comments?

File: src/jukebox/components/controls/common/evdev_listener.py [hunk 1]
```diff
@@ -49,14 +49,12 @@ def _filter_by_device_name(all_devices: List[evdev.InputDevice],
...
-
-        #. FileNotFoundError, if no device is found.
-        #. AttributeError, if device does not have the mandatory keys
+    :raise FileNotFoundError: if no device is found.
+    :raise AttributeError: if device does not have the mandatory key
 
     If multiple devices match, the first match is returned
 
-    :param device_name: See :func:`_filter_by_device_name`
+    :param device_name: See :func:`_filter_by_device_name
```
Comment (pabera) [inline src/jukebox/components/controls/common/evdev_listener.py:57]: Is this change correct?

Comment (s-martin) [reply src/jukebox/components/controls/common/evdev_listener.py:57]: Nope.

although the markdown formatter doesn't pick up the rst syntax, at least the formatting should be closed.

Will fix this.

File: documentation/developers/docstring/README.md [hunk 2]
```diff
@@ -0,0 +1,5913 @@
...
+
+
+
+<a id="components.gpio.gpioz.plugin.service_is_running_callbacks"></a>
+
+#### service\_is\_running\_callbacks
+
+Callback handler instance for service_is_running_callbacks events.
+
+See :class:`ServiceIsRunningCallbacks`
```
Comment (AlvinSchiller) [inline documentation/developers/docstring/README.md:3110]: Should `:class:` be rendered? Or be a link?

compare change in rfid/reader/__init__.py #55
https://github.com/s-martin/RPi-Jukebox-RFID/blob/63ecd190ea7303a7c643b79615f07b34a2f81a4e/src/jukebox/components/rfid/reader/__init__.py#L55

File: src/jukebox/components/rfid/hardware/template_new_reader/template_new_reader.py [hunk 3]
```diff
@@ -48,7 +48,7 @@ class ReaderClass(ReaderBaseClass):
     All the required interfaces are implemented there.
 
     Put your code into these functions (see below for more information)
-      - __init__
+      - `__init__`
```
Comment (AlvinSchiller) [inline src/jukebox/components/rfid/hardware/template_new_reader/template_new_reader.py:51]: Why is this changed?

File: documentation/developers/docstring/README.md [hunk 4]
```diff
@@ -0,0 +1,5913 @@
...
+
+The audio configuration of the system is one of those topics,
+which has a myriad of options and possibilities. Every system is different and PulseAudio unifies these and
+makes our life easier. Besides, it is only option to support Bluetooth at the moment.
+
+## Callbacks:
+
+The following callbacks are provided. Register callbacks with these adder functions (see their documentation for details):
+
+1. :func:`add_on_connect_callback`
```
Comment (AlvinSchiller) [inline documentation/developers/docstring/README.md:1268]: Should `:func:` be rendered?

File: src/jukebox/components/rfid/hardware/template_new_reader/template_new_reader.py (hunk 3 above)
Comment (s-martin) [reply src/jukebox/components/rfid/hardware/template_new_reader/template_new_reader.py:51]: Otherwise the double underscores would render as italic formatting

File: documentation/developers/docstring/README.md [hunk 5]
```diff
@@ -0,0 +1,5913 @@
...
+
+```python
+def register(func: Callable[[str, STATE], None])
+```
+
+Add a new callback function :attr:`func`.
+
+Callback signature is
+
+.. py:function:: func(folder: str, state: STATE)
```
[the remark this answers is not in the document]
Comment (s-martin) [reply documentation/developers/docstring/README.md:785]: The current [crossref processor](https://github.com/NiklasRosenstein/pydoc-markdown/blob/develop/src/pydoc_markdown/contrib/processors/crossref.py) seems to work only on classes. That’s why I left py:function and also :attr: as it is right now.

File: documentation/developers/docstring/README.md [hunk 6]
```diff
@@ -0,0 +1,5913 @@
...
+
+<a id="components.playermpd.playcontentcallback.PlayContentCallbacks.register"></a>
+
+#### register
+
+```python
+def register(func: Callable[[str, STATE], None])
+```
+
+Add a new callback function :attr:`func`.
```
[the remark this answers is not in the document]
Comment (s-martin) [reply documentation/developers/docstring/README.md:781]: See above

File: documentation/developers/docstring/README.md (hunk 2 above)
Comment (s-martin) [reply documentation/developers/docstring/README.md:3110]: Not all occurrences of :class: are currently converted to # syntax

File: documentation/developers/docstring/README.md (hunk 4 above)
Comment (s-martin) [reply documentation/developers/docstring/README.md:1268]: See above

File: documentation/developers/docstring/README.md [hunk 7]
```diff
@@ -0,0 +1,5913 @@
...
+
+Callback signature is
+
+.. py:function:: func(folder: str, state: STATE)
+    :noindex:
+
+    :param folder: relativ path to folder to play
+    :param state: indicator of the state inside the calling
+
```
[the remark this answers is not in the document]
Comment (s-martin) [reply documentation/developers/docstring/README.md]: done. not sure, why i cant resolve the conversation.

**Comment by AlvinSchiller** [comment]
Do we need an action that checks for changed but uncommited docs? As the precommit hook must be activated manually and could be forgotten (like flake8).

The action could run "./run_docgeneration.sh" and fail the build if there a outgoing changes?

**Comment by s-martin** [comment]
> Do we need an action that checks for changed but uncommited docs? As the precommit hook must be activated manually and could be forgotten (like flake8).
> 
> The action could run "./run_docgeneration.sh" and fail the build if there a outgoing changes?
> 
> 

I would suggest the following:
- keep the current pre commit hook, so devs can already check their changes
- in addition add a GitHub Action in case someone didn't activate pre commit hooks (similar to flake8 behavior)

My question would be, should I extend the Python action or add another one? What do you think?

Discussion about action **OR** hook was already in #2159

**Comment by AlvinSchiller** [comment]
Extending the current action would be fine for me.

Would you also add `flake8-docstrings` like mentioned in https://github.com/MiczFlor/RPi-Jukebox-RFID/pull/2159?

**Comment by s-martin** [comment]
I would add flake8-docstrings in a separate PR. Right now there are many warnings, which would fix in this separate PR.

**Comment by s-martin** [comment]
Added a check in the action. Needs more testing, though.

**Comment by s-martin** [comment]
The action does not work as expected, therefore I will add the integration into the Python action in a separate PR later.

Otherwise it could be merged from my point of view.

**Review by AlvinSchiller — APPROVED** [review]
